File: Java/Exceptions.java
package Java;

/**
 * Exceptions
 */
public class Exceptions {

    // user-defined custom exception, must extends some Exception object
    public static class MyException extends Exception {
        /**
         *
         */
        private static final long serialVersionUID = 1L;

        public MyException(String msg) {
            super(msg);
        }
    }

    public static void main(String[] args) {
        // can't throw any type
        // try {
        //     throw 10;
        // } catch (int e) {
        //     System.out.println(e);
        // }
        // handling built-in exceptions
        try {
            System.out.println(4/0);
        } catch (ArithmeticException e) {
            System.out.println(e.getMessage());
        }
        System.out.println("Program still running");
        // throw exception object
        try {
            throw new Exception("Exception");
        } catch (Exception e) {
            System.out.println(e.getMessage());
        }
        // finally runs whether execption is thrown or not
        int arr[] = {0, 1, 2};
        try {
            System.out.println(arr[1]);
        } catch (IndexOutOfBoundsException e) {
            System.out.println(e.getMessage());
        } finally {
            System.out.println("FINALLY");
        }
        try {
            System.out.println(arr[3]);
        } catch (IndexOutOfBoundsException e) {
            System.out.println(e.getMessage());
        } finally {
            System.out.println("FINALLY");
        }
        // using user-defined Exception
        try {
            throw new MyException("My Exception");
        } catch (MyException e) {
            System.out.println(e.getMessage());
        }
        // catch all by using Exception (base class), must be caught last or it will catch all exceptions (think of filtering system from top to bottom)
        try {
            throw new ArithmeticException("Arithmetic Exception");
        } catch (IndexOutOfBoundsException e) {
            System.out.println("Caught with IndexOutOfBounds: " + e.getMessage());
        } catch (Exception e) {
            System.out.println("Caught with Exception: " + e.getMessage());
        }
    }

}

File: Python/Exceptions.py
# built-in exceptions
try:
    print(2/0)
except ZeroDivisionError as e:
    print(e)

# else: runs only when exception is not raised in try block
# finally: always run
arr = [0, 1, 2]
try:
    print(arr[3])
except IndexError as e:
    print(e)
else:
    print("No Exception")
finally:
    print("Finally")

print()

try:
    print(arr[2])
except IndexError as e:
    print(e)
else:
    print("No Exception")
finally:
    print("Finally")

print()

# user-defined exception (raise keyword)
# inherit Exception base class
class NotEvenException(Exception):
    def __init__(self, msg=None):
        if msg is None:
            msg = "NotEvenException"
        self.msg = msg
    
    def __str__(self):
        return self.msg


def printEven(num):
    if (num % 2 != 0):
        raise NotEvenException("Given number is not even")
    print(num)

try:
    printEven(2)
    printEven(3)    
except NotEvenException as e:
    print(e)

